### PR TITLE
Create `flakey` decorator for retrying flaky tests and apply it to #143 and #148

### PR DESCRIPTION
**Pull request by jiridanek**

### CODE REVIEW

File: tests/system_test.py [hunk 1]
```diff
@@ -113,6 +114,39 @@ def _check_requirements():
...
+            # Python 3 clears the exception upon leaving the `except` block
+            # https://docs.python.org/3/reference/compound_stmts.html#the-try-statement
+            preserved_exc = None
+            for _ in range(repeats):
+                try:
+                    return f(*args, **kwargs)
+                except AssertionError as exc:
+                    preserved_exc = exc
+                    logging.info("Test marked as flaky has failed: %s", exc)
+            raise AssertionError("Flaky test has failed too many times") from preserved_exc
```
Comment (jiridanek) [inline tests/system_test.py:143]: Ken seems to use `flakey`, I am used to seeing `flaky`. I'm going to use the Ken's terminology thoughout.

**Comment by ganeshmurthy** [comment]
A test that is failing intermittently (a flakey test) either is badly written and needs to be fixed or it is it is exposing a bug in the router. I would rather fix the test bug or the router bug than marking the test as flakey.

There is usually already an issue associated with a flakey test. If a test is marked as flakey and say the test does not fail for a month or so, do we consider that issue fixed  and close it ? Or, are we going to come back to the issue and fix it when we have some time ?

File: tests/system_test.py [hunk 2]
```diff
@@ -113,6 +114,39 @@ def _check_requirements():
...
+        def inner(*args, **kwargs):
+            # Python 3 clears the exception upon leaving the `except` block
+            # https://docs.python.org/3/reference/compound_stmts.html#the-try-statement
+            preserved_exc = None
+            for _ in range(repeats):
+                try:
+                    return f(*args, **kwargs)
+                except AssertionError as exc:
+                    preserved_exc = exc
+                    logging.info("Test marked as flakey has failed: %s", exc)
```
Comment (jiridanek) [inline tests/system_test.py:142]: Maybe increase the log level of that.

**Comment by jiridanek** [comment]
all good points; I knew, going in, that `@flakey` is controversial

> If a test is marked as flakey and say the test does not fail for a month or so, do we consider that issue fixed and close it?

That's how it worked in Dispatch, I'd say; Cannot reproduce -> Closed

**Comment by jiridanek** [comment]
> mark a test as flakey and forget about it

We should not forget. But we should not be nagged by them all the time. I think the idea is to spend less time on CI review revisiting tests we know they are unreliable, again and again. Having the unreliable tests tracked in the Issues should be sufficient. Contributors should not be forced to investigate random flaky failures when they want to propose a PR.

Doing a `grep '@flakey' -A3` will quickly show what tests need fixing.

**Comment by jiridanek** [comment]
> The end result is we are going to ignore fixing the test which for me has the same effect as skipping a test.

Largely yes, but not quite. A `@flakey`-decorated test still has to be capable of passing at least sometimes, because the CI checks for that. So we know it is either flakey, or possibly fixed. It may never become completely broken (=incapable of passing).

Furthermore, the `@flakey` decorator only catches `AssertionError`, so if the test starts failing somehow unexpectedly, with some other exception, that will be also revealed in CI.

The router has been disabling tests in the past. Marking them as flakey would've been better solution, IMO. (Fixing is of course best, but that's matter of prioritization. The `@flakey` decorator gives you that flexibility of postponing the investigation/fix and still having usable CI that reveals problems in all the other reliable tests, and does not waste people's time having to deal with the flaky fails over and over.

(The last time I suggested doing something like this (autoretry) the folks did mostly dislike that; lets see what other people think now, ~3 years later)

File: tests/system_test.py [hunk 3]
```diff
@@ -114,6 +115,39 @@ def _check_requirements():
...
+    If applied, executes the test up to three times, marking it as failed only if it fails each time.
+    Note, that use of this decorator is generally discouraged---tests should pass reliably when their assertions are upheld.
+
+    Example usage
+
+        @flakey(issue='#123')
+        def test_feature_in_fragile_manner():
+            self.assertTrue(...)
+    """
+    del issue  # unused
```
Comment (kgiusti) [inline tests/system_test.py:130]: One thing to consider is to force the use of the issue parameter so we're certain there is a corresponding GH issue for the test.  Keeps us honest. And probably worth including the issue in the logging/assertion error text.

**Review by kgiusti — COMMENTED** [review]
It's actually "flaky" according to a quick dictionary lookup.  Not saying we should change it (unless you want to) but truth be known I'm a lousy speller...

Ganesh has a good point that this decorator can easily be abused by over-use.

I'm okay with this as long as we have solid evidence that the test is truly flaky and if so we should probably make fixing/removing them release blockers.

But I'll delegate approval to Mr. Murthy since he's the CI Czar...

**Comment by jiridanek** [comment]
Closed due to the lack of enthusiasm.

**Pull request closed without merging**